### PR TITLE
MDEV-18463: Don't allow multiple table CONSTRAINTs with the same name

### PR DESCRIPTION
**Pull request by an3l**

### CODE REVIEW

File: sql/sql_table.cc [hunk 1]
```diff
@@ -4203,6 +4203,20 @@ mysql_prepare_create_table(THD *thd, HA_CREATE_INFO *create_info,
...
         }
+
+        /* Check that there's no repeating foreign key constraint names. */
+        List_iterator_fast<Key> key_it(alter_info->key_list);
+        Key *key;
+        while ((key= key_it++))
+        {
+          if (check->name.length == key->name.length &&
+              my_strcasecmp(system_charset_info,
+                            check->name.str, key->name.str) == 0 && key->type == Key::FOREIGN_KEY)
```
Comment (dr-m) [inline sql/sql_table.cc]: InnoDB may already check this indirectly by defining a unique index on `SYS_FOREIGN.ID`. But, InnoDB does not store any `CHECK` constraints. And we’d like to move the FOREIGN KEY constraint metadata storage out of InnoDB.

I think that constraint names should be unique, no matter what the constraint type is.
Is the condition `&& key->type == KEY::FOREIGN_KEY` really necessary?

Comment (an3l) [reply sql/sql_table.cc]: Changed thanks @dr-m

File: sql/sql_table.cc [hunk 2]
```diff
@@ -4203,6 +4203,20 @@ mysql_prepare_create_table(THD *thd, HA_CREATE_INFO *create_info,
             DBUG_RETURN(TRUE);
           }
         }
+
+        /* Check that there's no repeating foreign key constraint names. */
+        List_iterator_fast<Key> key_it(alter_info->key_list);
+        Key *key;
+        while ((key= key_it++))
```
Comment (dr-m) [inline sql/sql_table.cc]: I do not see a check against duplicated `FOREIGN KEY` constraint names. I do see a check against the same name being used for both `CHECK` and `FOREIGN KEY` constraints. Can you revise the comment?
I would suggest introducing `const` and reducing the scope of a variable, like C++98 allows us to do:
```c++
while (const Key* key= key_it++)
```

File: mysql-test/t/check_constraint.test [hunk 3]
```diff
@@ -176,3 +176,16 @@ select * from t1 where a is null;
...
+--echo #
+
+create table tpk
+(id INT NOT NULL AUTO_INCREMENT PRIMARY KEY,name VARCHAR(100) NOT NULL);
+--error ER_DUP_CONSTRAINT_NAME
+create table tfk
+(id INT, c1 INT, c2 INT NOT NULL,
+ CONSTRAINT sid FOREIGN KEY (`c1`) REFERENCES tpk (`id`)
+ ON DELETE NO ACTION ON UPDATE NO ACTION, CONSTRAINT sid check (c2>15));
+drop table tpk;
```
Comment (dr-m) [inline mysql-test/t/check_constraint.test]: I would suggest more tests, and I think that we need a variant of the test where both tables are created with `ENGINE=InnoDB` in order to actually have the `FOREIGN KEY` constraints created:
```sql
CREATE TABLE tpk(id INT PRIMARY KEY);
CREATE TABLE tfk (c1 INT, c2 INT, CONSTRAINT sid FOREIGN KEY (c1) REFERENCES tpk (id));
--error ER_DUP_CONSTRAINT_NAME
ALTER TABLE tfk ADD CONSTRAINT Sid CHECK (c2>15);
ALTER TABLE tfk DROP FOREIGN KEY sid;
ALTER TABLE tfk ADD CONSTRAINT Sid CHECK (c2>15);
--error ER_DUP_CONSTRAINT_NAME
ALTER TABLE tfk ADD CONSTRAINT sid FOREIGN KEY (c1) REFERENCES tpk (id);
--error ER_DUP_CONSTRAINT_NAME
ALTER TABLE tfk ADD CONSTRAINT FOREIGN KEY sid (c1) REFERENCES tpk (id);
ALTER TABLE tfk ADD CONSTRAINT tid FOREIGN KEY (c1) REFERENCES tpk (id);
DROP TABLE tfk, tpk;
```
In `main.foreign_key`, I see no `ER_DUP_CONSTRAINT_NAME`. I would suggest to add a test for that too:
```sql
CREATE TABLE tfk (c1 INT, c2 INT, CONSTRAINT sid FOREIGN KEY (c1) REFERENCES tpk (id));
--error ER_DUP_CONSTRAINT_NAME
ALTER TABLE tfk ADD CONSTRAINT sid FOREIGN KEY (c2) REFERENCES tpk(id);
DROP TABLE tfk,tpk;
```

Comment (an3l) [reply mysql-test/t/check_constraint.test]: Hi @dr-m,I have updated tests.

The last test case is showing the field check constraint vs foreign key constraint with the same name. Currently it is not covered so I'm not sure should be covered at all. 
Please let me know should this be changed also.
Thanks

File: mysql-test/t/check_constraint.test [hunk 4]
```diff
@@ -176,3 +161,13 @@ select * from t1 where a is null;
...
+
+--echo #
+--echo # MDEV-18463 Don't allow multiple table CONSTRAINTs with the same name
+--echo #
+
+let $engine = MyISAM;
+source include/forbid_same_constraint_name_mdev18463.inc;
+
+let $engine = InnoDB;
+source include/forbid_same_constraint_name_mdev18463.inc;
```
Comment (dr-m) [inline mysql-test/t/check_constraint.test]: Do not omit the line terminator.
Do we really need `eval` and `$engine` here?
```sql
SET @save_engine = @@SESSION.default_storage_engine;
SET default_storage_engine=innodb;
source include/forbid_same_constraint_name_mdev18463.inc;
SET default_storage_engine=@save_engine;
source include/forbid_same_constraint_name_mdev18463.inc;
```

File: mysql-test/include/forbid_same_constraint_name_mdev18463.inc [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+# This doesn't drop index KEY('sid')
+ALTER TABLE tfk DROP FOREIGN KEY sid;
+SHOW CREATE TABLE tfk;
+
+--echo #
+--echo # Drop the foreign key will not drop the index so adding new constraint
+--echo # will raise an error
+--echo #
+--error ER_DUP_CONSTRAINT_NAME
+ALTER TABLE tfk ADD CONSTRAINT Sid CHECK (c2>15);
```
Comment (dr-m) [inline mysql-test/include/forbid_same_constraint_name_mdev18463.inc]: Is a non-unique index really a constraint? If `ADD CONSTRAINT…FOREIGN KEY` is able to add implicit `KEY` with the same name, isn’t that a proof that index names and constraint names are disjoint domains?

I might accept that `UNIQUE` index counts as a `CONSTRAINT` name, but what does the standard say?

File: mysql-test/include/forbid_same_constraint_name_mdev18463.inc [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+ALTER TABLE tfk ADD CONSTRAINT sid FOREIGN KEY (c1) REFERENCES tpk (id);
+SHOW CREATE TABLE tfk;
+SELECT constraint_name, check_clause FROM information_schema.check_constraints;
+
+DROP TABLE tfk;
+--echo #
+--echo # Create table with unnamed table check constraint and foreign constraint with the same name
+--echo #
+--error ER_DUP_CONSTRAINT_NAME
+eval CREATE TABLE tfk (c1 INT, c2 INT, CONSTRAINT CONSTRAINT_1 FOREIGN KEY (c1)                        REFERENCES tpk (id), CONSTRAINT CHECK (c2>0))
```
Comment (dr-m) [inline mysql-test/include/forbid_same_constraint_name_mdev18463.inc]: It looks like a line break is missing here.

File: mysql-test/include/forbid_same_constraint_name_mdev18463.inc [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+                       ENGINE=$engine;
+
+--echo #
+--echo # Field check constraints name same as name of field `c2` and as foreign key
+--echo #
+--error ER_DUP_CONSTRAINT_NAME
+eval CREATE TABLE tfk (c1 INT, c2 INT check (c2>0), CONSTRAINT c2
+                       FOREIGN KEY (c1) REFERENCES tpk (id)) ENGINE=$engine;
+
+DROP TABLE tpk;
```
Comment (dr-m) [inline mysql-test/include/forbid_same_constraint_name_mdev18463.inc]: Do not omit the trailing line terminator.

File: mysql-test/include/forbid_same_constraint_name_mdev18463.inc (hunk 5 above)
Comment (holyfoot) [reply mysql-test/include/forbid_same_constraint_name_mdev18463.inc]: AFAIK the SQL standard doesn't mention non-unique index constraint.

**Comment by an3l** [comment]
Pushed with fd1979bc9a53 to `10.2`.

**Pull request closed without merging**